core/src/main/java/com/linecorp/armeria/common/multipart/MultipartDecoder.java

```diff
@@ -0,0 +1,312 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.linecorp.armeria.common.multipart;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.LinkedList;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+
+import javax.annotation.Nullable;
+
+import org.reactivestreams.Processor;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import com.linecorp.armeria.common.HttpData;
+import com.linecorp.armeria.common.HttpHeaders;
+import com.linecorp.armeria.common.HttpHeadersBuilder;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufHolder;
+import io.netty.buffer.Unpooled;
+import io.netty.util.ReferenceCountUtil;
+
+/**
+ * Reactive processor that decodes HTTP payload as a stream of {@link BodyPart}.
+ */
+class MultipartDecoder implements Processor<HttpData, BodyPart> {
+
+    // Forked from https://github.com/oracle/helidon/blob/a9363a3d226a3154e2fb99abe230239758504436/media
+    // /multipart/src/main/java/io/helidon/media/multipart/MultiPartDecoder.java
+
+    private static final AtomicReferenceFieldUpdater<MultipartDecoder, Subscription> upstreamUpdater =
+            AtomicReferenceFieldUpdater.newUpdater(
+                    MultipartDecoder.class, Subscription.class, "upstream");
+
+    @SuppressWarnings("rawtypes")
+    private static final AtomicReferenceFieldUpdater<MultipartDecoder, Subscriber> downstreamUpdater =
+            AtomicReferenceFieldUpdater.newUpdater(
+                    MultipartDecoder.class, Subscriber.class, "downstream");
+
+    private final CompletableFuture<BufferedEmittingPublisher<BodyPart>> initFuture;
+    private final LinkedList<BodyPart> bodyParts;
+    private final MimeParser parser;
+    private final ParserEventProcessor parserEventProcessor;
+
+    @Nullable
+    private BufferedEmittingPublisher<BodyPart> emitter;
+    @Nullable
+    private BodyPartBuilder bodyPartBuilder;
+    @Nullable
+    private HttpHeadersBuilder bodyPartHeaderBuilder;
+    @Nullable
+    private BufferedEmittingPublisher<HttpData> bodyPartPublisher;
+
+    // Updated only via upstreamUpdater
+    @Nullable
+    private volatile Subscription upstream;
+    // Updated only via downstreamUpdater
+    @Nullable
+    private volatile Subscriber<? super BodyPart> downstream;
+
+    MultipartDecoder(String boundary) {
+        requireNonNull(boundary, "boundary");
+        parserEventProcessor = new ParserEventProcessor();
+        parser = new MimeParser(boundary, parserEventProcessor);
+        initFuture = new CompletableFuture<>();
+        bodyParts = new LinkedList<>();
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super BodyPart> subscriber) {
+        requireNonNull(subscriber, "subscriber");
+        if (emitter != null || !downstreamUpdater.compareAndSet(this, null, subscriber)) {
```

Comment:
Probably we don't need to check `emitter != null `.

Reply:
That's true. However we can reduce to call CAS operation by checking `emitter != null`.